.github/workflows/appsec.yml

```
@@ -26,168 +26,230 @@ on:
+        runs-on: [ macos-11, macos-14 ] # oldest and newest macos runners available - macos-14 mainly is here to cover the fact it an ARM machine
         go-version: [ "1.21", "1.20", "1.19" ]
-        cgo_enabled: [ "0", "1" ] # test it compiles with and without cgo
-        appsec_enabled: # test it compiles with and without appsec enabled
-          - DD_APPSEC_ENABLED=true
-          - DD_APPSEC_ENABLED=false
-          - "" # the env var is not defined so that the remote-config path can be taken
-        include:
-          - cgocheck:
-              GODEBUG=cgocheck=2
-          - go-version: "1.21"
-            cgocheck:  # 1.21 deprecates the GODEBUG=cgocheck=2 value, replacing it with GOEXPERIMENT=cgocheck2
-              GOEXPERIMENT=cgocheck2
-      fail-fast: false
-    name: native ${{ toJSON(matrix) }}
-    runs-on: ${{ matrix.runs-on }}
+      fail-fast: true # saving some CI time - macos runners too long to get
     steps:
       - uses: actions/checkout@v4
+
+      - name: Restore Go modules cache
+        uses: actions/cache/restore@v4
         with:
-          ref: ${{ inputs.ref || github.ref }}
+          path: ${{ needs.go-mod-caching.outputs.path }}
+          key: ${{ needs.go-mod-caching.outputs.key }}
+          restore-keys: go-pkg-mod-
+          enableCrossOsArchive: true
+          fail-on-cache-miss: true
+
       - uses: actions/setup-go@v5
         with:
           go-version: ${{ matrix.go-version }}
+          cache: false # we manage the caching ourselves
 
+      # go test is being manually called multiple times here for the sake of reusing the runner.
+      # Waiting runners is unfortunately so long that we decided to do so for things only requiring recompilation or
+      # reruns under different settings.
       - name: go test
         shell: bash
+        env:
+          GOMODCACHE: ${{ github.workspace }}/${{ needs.go-mod-caching.outputs.path }}
         run: |
-          go install gotest.tools/gotestsum@latest
-          # Run the tests with gotestsum
-          env ${{ matrix.cgocheck }} CGO_ENABLED=${{ matrix.cgo_enabled }} ${{ matrix.appsec_enabled }} ./.github/workflows/apps/appsec-test-contrib-submodules.sh
-
-      - name: Upload the results to Datadog CI App
-        uses: ./.github/actions/dd-ci-upload
-        with:
-          dd-api-key: ${{ secrets.DD_CI_API_KEY }}
-          files: ${{ env.JUNIT_REPORT }}*.xml
-          tags: go:${{ matrix.go-version }},arch:${{ runner.arch }},os:${{ runner.os }}
+          set -ex
+          cgocheck="GOEXPERIMENT=cgocheck2"
+          if [[ "$(go version)" =~ (go1.20)|(go1.19) ]]; then
+            cgocheck="GODEBUG=cgocheck=2"
+          fi
+          for cgo in 0 1; do
+            for appsec_enabled_env in "" "DD_APPSEC_ENABLED=true" "DD_APPSEC_ENABLED=false"; do
+                for cgocheck_env in "" "$cgocheck"; do
+                  if ! env CGO_ENABLED=$cgo $appsec_enabled_env $cgocheck_env go test -v $TESTS; then
```

```suggestion
                  if ! env CGO_ENABLED=$cgo $appsec_enabled_env $cgocheck_env go test -v -shuffle=on -parrallel=2 -count=4 $TESTS; then
```
Probably useful and won't add to much time to the CI but can make sure cleanup and parallelism is handled correctly

---

To do in a separate PR, as I believe it doesn't pass